supervisor.go

```diff
@@ -133,30 +139,65 @@ func (s *Supervisor) Setargs(ctx context.Context, args []string, _ *struct{}) er
 // is appended to the default process environment: keys provided here
 // override those that already exist in the environment.
 func (s *Supervisor) Setenv(ctx context.Context, env []string, _ *struct{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	s.environ = env
 	return nil
 }
 
-// Exec reads a new image from its argument and replaces the current
-// process with it. As a consequence, the currently running machine will
-// die. It is up to the caller to manage this interaction.
-func (s *Supervisor) Exec(ctx context.Context, exec io.Reader, _ *struct{}) error {
+// Setbinary uploads a new binary to replace the current binary when
+// Supervisor.Exec is called. The two calls are separated so that
+// different timeouts can be applied to upload and exec.
+func (s *Supervisor) Setbinary(ctx context.Context, binary io.Reader, _ *struct{}) error {
 	f, err := ioutil.TempFile("", "")
 	if err != nil {
 		return err
 	}
-	if _, err := io.Copy(f, exec); err != nil {
+	if _, err := io.Copy(f, binary); err != nil {
 		return err
 	}
 	path := f.Name()
 	if err := f.Close(); err != nil {
+		os.Remove(path)
 		return err
 	}
 	if err := os.Chmod(path, 0755); err != nil {
+		os.Remove(path)
 		return err
 	}
+	s.mu.Lock()
+	s.binaryPath = path
+	s.mu.Unlock()
+	return nil
+}
+
+// Getbinary retrieves the last binary uploaded via Setbinary.
+func (s *Supervisor) Getbinary(ctx context.Context, _ struct{}, rc *io.ReadCloser) error {
```

Comment:
Is this used anywhere?

Reply:
not yet, but might as well implement it since we can. (i plan on using this later.)